# HW 01 
# Gary Fowler


## Exercise 2.1 

### Another ball dropped from a tower


A ball is dropped from a tower of height h, with initial velocity zero. Write a program that asks for a height and then
calculates the time it takes for the ball to hit the ground, ignoring air resistance.

In [1]:
## @author: Gary Fowler

"""
@h: height
@a: acceleration due to gravity
@formula: this is the formula s = .5*a*t^2 solved for t, position is replaced 
    with tower height

in this program the variable needed is time, height is our position

"""
import numpy as np

h = float(input("Enter the tower height: "))
a = 9.81 

#this formula returns seconds, it is derived from position = .5*a*t^2
formula = np.sqrt(2*h/a)

#printing out the result and rounding to 3 digits
print("The ball will hit the ground after {number:.{digits}f} seconds.".format(number=formula, digits=3))


Enter the tower height: 100
The ball will hit the ground after 4.515 seconds.


## Exercise 2.3

### Cartesian coordinates to polar coordinates


Prompt user for Cartesian coordinates (x,y) of a point in 2d space, calculate and print the polar coordinates, with 
the angle theta in degrees.

In [2]:
## @author: Gary Fowler

"""
@x: the x term of 2d point
@y: the y term of 2d point
@r: the radius in polar coordinates
@thetaRadians: this is the solution to atan2(x,y) which is in radians
@thetaDegrees: this is the conversion of radians to degrees

in this program we are converting x,y to r,theta

"""
import numpy as np
import math as m

x = int(input("Enter x of the Cartesian coordinates (x,y) for a point in 2d space: "))
y = int(input("Enter y of the Cartesian coordinates (x,y) for a point in 2d space: "))

## r is sqrt(x^2 + y^2)
r = np.sqrt(x**2 + y**2)
## theta is found in radians, must be converted to degrees using 180/pi
thetaRadians = m.atan2(x, y)
thetaDegrees = thetaRadians * 180 / np.pi

## again we are formatting our output to 2 decimals
print("r = {number:.{digits}f} ".format(number=r, digits=3))
print("theta =", thetaDegrees,"degrees")

Enter x of the Cartesian coordinates (x,y) for a point in 2d space: -1
Enter y of the Cartesian coordinates (x,y) for a point in 2d space: -1
r = 1.414 
theta = -135.0 degrees


## Exercise 2.5

### Quantum potential step


A particle of mass m that encounters a one-dimensional potential step. The particle encounters a sudden jump in potential energy of height V at position x = 0. It can either a) pass the step or b) be reflected.
The particle we are interested in has mass m = 9.11 x 10^-31kg and energy 10 eV.

In [3]:
## @author: Gary Fowler

"""
@m: the mass of our particle
@E: the Energy of our particle
@h: the reduced Planck's constant
@V: the potential energy step height
@k1: the initial wavevector
@k2: smaller second wavevector after the potential energy step
@transmission: formula for if the particle passes the step
@relfection: formula for if the particle is reflected

in this program we will print out the transmission and reflection probabilities using given formulas and values

"""
import numpy as np

m = 9.11e-31
E = 10 ##units are eV
V = 9 ##the height of the potential step
h = (4.1356e-15) / 2 * np.pi ##the reduce Planck's constant

k1 = np.sqrt(2*m*E) / h
k2 = np.sqrt(2*m*(E - V)) / h

transmission = (4*k1*k2) / ((k1 + k2)**2)
reflection = ((k1 - k2) / (k1 + k2))**2

##we are not going to be formatting this output as the precision may be necessary
##since I do not know quantum physics I am unsure if it is necessary or not
print("If the particle is Transmitted: ",transmission)
print("Or if the particle is Reflected: ",reflection)

If the particle is Transmitted:  0.730126136388
Or if the particle is Reflected:  0.269873863612


## Exercise 2.6

### Planetary orbits


In general orbits are elliptical. If given the distance L1 of closest approach that a planet makes to the Sun, also called is perihelion, and its linear velocity v1 at perihelion, then any other property of orbit can be calculated from these two:

a) Kepler's second law, L2 and v2 at the most distant point, aphelion, satsify L2v2 = L1v1. The total energy is E = 1/2mv^2 - G(mM / r). Energy must be conserved so show that v2 is the smaller root of:
v2^2 - (2GM / v1L1)v2 - [v1^2 - (2GM /L1)] = 0. Then once we have v2 we can calculate L2.

b) Given values v1, L1, and L2, other parameters of the orbit are given by simple formulas from Kepler's laws:
semi-major axis: a = 1/2(L1 + L2)
semi-minor axis: b = sqrt(L1L2)
orbital period: T = 2piab / L1v1
orbital eccentricity: e = (L2 - L1 / L2 + L1)
create the program to ask the user for distance to the Sun and veloctiy at perihelion, then calculate and print L2, v2, T, and e

c) Test the program by having it calculate the properties of Earth and Halley's comet. Earth orbital period is 1, Halley's comit is 76 years.

In [19]:
## @author: Gary Fowler

"""
block a)
@m: the mass of the planet
@E: the total Energy, kinetic plus gravitational, of a planet distance r from Sun
@M: the mass of the Sun
@L1: the length at the perihelion
@L2: the length at the aphelion
@v1: the velocity at perihelion
@v2: the velocity at aphelion
@a,b,c: used for the quadratic formula
@perihelionVelocity: this verifies v1, and does not override the given v1
@aphelionVelocity: this is solved from v1, and be reassigned v2 after output

if a variable is used again from a, we will not be redefining it in b or c


in this first block of the program we will use Kepler's second law and a given L1 and v1 at perihelion to determine L2 and v2
as well as verify that v2 is the smaller root in order to conserve energy.

"""

import numpy as np

L1 = float(input("Enter the perihelion: "))
v1 = float(input("Enter the veloctiy at the perihelion: "))

##Earth's properties
#L1 = 1.4710e11 
#v1 = 3.0287e4

##Halley's comet properties
#L1 = 8.7830e10
#v1 = 5.4529e4

##Newton's gravitational constant m^3 / kg s^2
G = 6.6738e-11
##Sun's mass in kg
M = 1.981e30

a = 1
b = -(2*G*M) / (v1*L1)
c = -(v1**2 - (2*G*M) / L1)

perihelionVelocity = (-b + np.sqrt(b**2 - 4*a*c))/ 2*a
##smaller than perihelion
aphelionVelocity = (-b - np.sqrt(b**2 - 4*a*c))/ 2*a

#print("The perihelion velocity:",perihelionVelocity)
print("The aphelion velocity:",aphelionVelocity)

##for ease of equations we will re-assign v2 as the aphelion
v2 = aphelionVelocity

L2 = (L1*v1) / v2
print("Thus the length at the aphelion is:",L2)

"""
block b)
if a variable is used again from a, we will not be redefining it in b or c
@semi_majorAxis: the formula to find the semi-major axis
@semi_minorAxis: the formula to find the semi_minor axis
@orbitalPeriod: the formula to find the orbital period

in this block we now have v1, L1, and L2, and will be finding other parameters of orbit

"""
semi_majorAxis = .5 * (L1 + L2)
#print("Semi-major axis:",semi_majorAxis)

semi_minorAxis = np.sqrt(L1 * L2)
#print("Semi-minor axis:",semi_minorAxis)

orbitalPeriod = (2 * np.pi * semi_majorAxis * semi_minorAxis) / (L1 * v1)
##seconds per day = 86400, convert seconds to years
convertYears = orbitalPeriod / (86400 * 365)
print("Orbital period:",convertYears)

orbitalEcc = (L2 - L1) / (L2 + L1)
print("Orbital eccentricity:",orbitalEcc)




Enter the perihelion: 8.7830e10
Enter the veloctiy at the perihelion: 5.4529e4
The aphelion velocity: 680.935880849
Thus the length at the aphelion is: 7.03338185679e+12
Orbital period: 116.420724
Orbital eccentricity: 0.975332850148


In the above example when inputting Earth's properties (L1 = 1.4710e11 and v1 = 3.0287e4) we get: 
Aphelion velocity: 29062.73
Aphelion length: 1.5329e11
Period: 1.009
Eccentricity: 0.0206

When inputting Halley's comet we receive
Aphelion veloctiy: 680.936
Aphelion length: 7.033e12
Period: 116 years
Eccentricity: 0.975


## Exercise 2.13

### Recursion


The Catalan numbers definition can be rewritten as Cn = {1 if n = 0, (4n-2)/(n+1) * Cn-1 if n > 0}
a)Use this to calculate and print C100
b)Euclid showed that the greatest common divisor g(m,n) of two nonnegative integers satisfies 
g(m,n) = {m if n = 0, g(n,m mod n) if n > 0}
write a funtion that employs recursion to calculate the gcd of 108 and 192

In [5]:
## @author: Gary Fowler

"""
block a)
@x: the xth catalan number we are looking for
@n: the parameter for the function catalans

@catalans(n): a function that takes one parameter,n, and returns the nth catalan number

this block will use recursion to print out the 100th Catalan number

"""
x = 100

def catalans(n): 
    if n==0:
        return 1
    else:
        return (4*n-2)*catalans(n-1) // (n+1) 
    
print("The",x,"th catalan number is:",catalans(x))

"""
block b)
@y, z: the numbers we want to find the gcd of
@m, n: the parameters for the function gcd(m,n)

@gcd(m,n): a function that takes two parameters and returns the gcd of the two

this block will use recursion to print out the gcd of 108 and 192

"""
y = 64
z = 729

def gcd(m,n):
    if n==0:
        return m
    else:
        return gcd(n,m % n)

print("The greatest common divisor between",y,"and",z,"is:",gcd(y,z))

The 100 th catalan number is: 896519947090131496687170070074100632420837521538745909320
The greatest common divisor between 64 and 729 is: 1
